### PR TITLE
Fix Issue #206 "Some Tumblr blogs are falsely reported as offline"

### LINKED ISSUE #206: Some Tumblr blogs are falsely reported as offline
**Describe the bug**
When a blog changes its state to hidden (dashboard blog), Tumblthree shows this blog as offline, even it can still be viewed in the web browser.

**To Reproduce**
1. select a offline blog that is now a hidden blog
2. click 'check blog online status' in the context menu
3. after the check the blog still shows as offline

**Expected behavior**
At least when doing a manual "check online status", it should be recognized that the blog is still available, but now needs to be accessed as hidden blog through the dashboard.
Tumblthree should make a double check, if necessary, and automatically change the blog type to hidden blog, if possible.

**Screenshots**

**Desktop (please complete the following information):**
 - TumblThree version: 2.4.8.243
 - OS: Windows 10 Pro
 - Browser: Chrome
 - Version: current

**Additional context**
At the moment a workaround is to go to the blog settings and change the crawler type to "SVC" for blogs that changed their state to hidden.

### PR DESCRIPTION
**Pull request by kulesy**
## Title

**Fixing:** Listed blogs in manager view that have subsequently been switched to hidden on Tumblr will appear - upon status check - as offline.

## Description

**How this has been fixed:** 
When executing a status check, the URL of the selected blog will be recognized to have a hidden route which will trigger the blog type to be tmblrpriv and the URL to have a hidden route while the status of the blog is still online. 
With this new URL, a get request is sent to the API which, if the user is logged in, will return 200 OK and the app runs normally. 
However if the get request returns a 404 NotFound an exception will be thrown and caught. Knowing that the blog is hidden the most likely reason for the exception will be because the user is not logged in, therefore an the "must be logged in" error message is displayed.

## Issue Resolution

This Pull Request Fixes #206 

## Proposed Changes

 - ITumblrBlogDetector to be injected into the TumblrBlogCrawler - to identify blog as hidden
 - GetUrlRedirection() in TumblrBlogDetector to be changed from private to public - to get the hidden blog URL

### CODE REVIEW

**Comment by thomas694** [comment]
Let's look at the code. Ok, from your questions and our descriptions, I had a different picture in my head. No problem, my imagination.

Is `IsBlogOnlineAsync` only called when doing a manual check initiated through the context menu or also internally while processing and so on? I'm not sure about side effects and would only implement it for a manual check. Is it possible to add an unchecked in the meantime hidden blog to the queue and crawl?
Looking inside a tmblrpriv file, I think normally the change of the URL wouldn't be necessary, as the API is still accessed with the normal URL, or it's reconstructed somewhere in the code.
It introduces a new dependency (`TumblrBlogDetector`), ok. But then the tmblr crawler changes the type of its assigned blog and checks a tmblrpriv blog with his own code. I haven't run it. That might work at the moment, but is a mixture of concerns. It's not so good in my opinion. One crawler only knows how to handle its own type.
I think, it would be better to do the decision making and blog (type) changes outside where the call to `IsBlogOnlineAsync` is done, inside e.g. the `ManagerController` you mentioned before.

What do you think about it? Wouldn't this approach be cleaner?

**Comment by kulesy** [comment]
Yes that looks to be a much better way of doing it, I should've looked a further back in the scope of status checks, my bad. Could you explain further what you mean by an unchecked for hidden blogs to be crawled. Do you mean while the online status is displayed and the blog is hidden it should still be able to be crawled?

**Comment by kulesy** [comment]
An online check is done within the the crawling controller (when the crawl is initiated), however the online status is checked within the type specific ICrawler class (eg. TumblrBlogCrawler) through the IsBlogOnlineAsync() which is where I had the hidden check originally. So I believe the check might actually have to be done through the TumblrBlogCrawler class and change the blog type there, which should be fine as a new instance of a TumblrHiddenCrawler class will be made for the actual crawling process.

**Comment by thomas694** [comment]
Yes, you're right in terms of flow where it's going through. But interface wise the individual crawler doesn't need to know, whether there is another crawler that can process that (changed) type and how, or try to do it itself.
After changing the type it's going on with `await GetApiPageWithRetryAsync(0)`, which seems to work at the moment, because internally nearly the same thing is done. But somehow it's mingling a bit. Also I'm thinking about the change of the URL, which you most probably need for continuing in the current crawler, but maybe makes more trouble in the future, because that changes more often than the standardized blog (domain) name.
Maybe you can have a look, whether it's possible to move this logic out of the crawler to the outside. If, after thinking about it or maybe a small try, you come to the conclusion that it gets too cumbersome to do it from outside, we can stick with your first way.

**Comment by kulesy** [comment]
Okay I've got the hidden blog type updating outside the crawlers now. However it required the crawler controller to be dependent on the TumblrBlogDetector. So now the hidden blog detection works for both the status check and the crawling of queued blogs. Therefore I believe you don't have to worry about any bugs occuring with incorrect blog types in specific crawlers. It also addresses if the user is signed in for hidden blogs without the need of the URL change which was a bit clunky.

**Review by thomas694 — CHANGES_REQUESTED** [review]
It looks good now. 🙂 
Only a small tiny bit of cleanup and it's ok.

File: src/TumblThree/TumblThree.Applications/Controllers/CrawlerController.cs [hunk 1]
```diff
@@ -205,7 +208,9 @@ private async Task RunCrawlerTasksAsync(PauseToken pt, CancellationToken ct)
                             continue;
                         }
                         IBlog blog = nextQueueItem.Blog;
-
+                        bool isHiddenTumblrBlog = await _tumblrBlogDetector.IsHiddenTumblrBlogAsync(blog.Url);
```
[the remark this answers is not in the document]
Comment (kulesy) [reply src/TumblThree/TumblThree.Applications/Controllers/CrawlerController.cs:211]: Yes that would make more sense.

File: src/TumblThree/TumblThree.Applications/Crawler/ITumblrBlogDetector.cs [hunk 2]
```diff
@@ -11,5 +11,7 @@ public interface ITumblrBlogDetector
         Task<bool> IsTumblrBlogAsync(string url);
 
         Task<bool> IsTumblrBlogWithCustomDomainAsync(string url);
+
+        Task<string> GetUrlRedirection(string url);
```
[the remark this answers is not in the document]
Comment (kulesy) [reply src/TumblThree/TumblThree.Applications/Crawler/ITumblrBlogDetector.cs:15]: I should've gone back in the PR to get rid of the unnecessary changes, apologies. Teaching me a lot, thanks.

**Comment by kulesy** [comment]
I'm assuming these little commits will be cleaned up in the merging process?

**Review by thomas694 — APPROVED** [review]
Yes, I'll do a squash merge.

**Comment by thomas694** [comment]
Thank you for your PR and contribution to the TumblThree project.
I'd be glad, if you like to solve the one or other issue in the future.

**Comment by kulesy** [comment]
I'd be happy to help out some more.

**Pull request merged**